### PR TITLE
Lift The "Large Space Heuristic" For Real This Time

### PR DESCRIPTION
**Pull request by CodaFi**
<strike>In light of the invocation limits placed on space subtraction, this grossly incorrect check is being dropped.  But, this re-opens an old wound: What to do about pattern matrices that are large enough to trip the first size check, but have sufficiently wild structure (due to constructor patterns under wildcards) which trip this heuristic and didn't diagnose before?  In the most severe cases, we would miscompile (see SR-6316).  Today, the minus limit just so happens to catch most of these patterns and diagnoses them.  So, we'll let that do the hard work and raise the maximum space size to compensate for as much of the middle distance as possible.

That said, this patch is narrowly source breaking, but for a class of programs that are simply unreasonable for humans to be writing, or have undefined semantics under the current compiler.</strike>

Resolves [SR-6316](https://bugs.swift.org/browse/SR-6316) and [SR-6652](https://bugs.swift.org/browse/SR-6652)

### CODE REVIEW

**Comment by CodaFi** [comment]
This old dumb heuristic is responsible for *all three* of the miscompiles between those two JIRA tickets.

**Comment by jrose-apple** [comment]
I haven't read the code yet but I'll mention that humans are usually not writing the terrible switch statements; code generators are.

(There are also reasonable ones, like an enum-of-ten-cases that contains an enum-of-ten-cases in most of the payloads.)

**Comment by jrose-apple** [comment]
Another good test case: something that holds a handful of POSIXError.Codes.

**Comment by CodaFi** [comment]
Lost my old test results, so I'm going to experiment with dropping the entire size heuristic.

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
@jrose-apple New plan: Drop this heuristic like a stone.  I'm flushing it, and the Sema regression tests but leaving in the compatibility tests for now.  Under the new heuristic, having a large amount of enum cases does not preclude you from coverage checking, producing a sufficiently convoluted matrix that causes the coverage checker to barf does.

We get the best of both worlds here: Source compatibility is retained and the compiler is now coverage-checking patterns it used to call "too complex" a la [SR-6316](
https://bugs.swift.org/browse/SR-6316).

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
⛵️

**Comment by jrose-apple** [comment]
Should we have some kind of manual timeout like constraint-solving does? That's why this went in in the first place.

File: lib/Sema/TypeCheckSwitchStmt.cpp [hunk 1]
```diff
@@ -1058,18 +949,11 @@ namespace {
...
-          estimateSearchSpaceSize(totalSpace, coveredSpace);
-      if (searchSpaceSizeEstimate > Space::getMaximumSize()) {
-        diagnoseCannotCheck(sawRedundantPattern, totalSpace, coveredSpace,
-                            unknownCase);
-        return;
-      }
       unsigned minusCount = 0;
       auto diff = totalSpace.minus(coveredSpace, TC, DC, &minusCount);
       if (!diff) {
-        diagnoseCannotCheck(sawRedundantPattern, totalSpace, coveredSpace,
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckSwitchStmt.cpp]: What happened to `sawRedundantPattern` here?

**Comment by davidungar** [comment]
I'm glad to see `minusCount` is still in there as a last-resort way to abort exponential compilation time.

**Comment by jrose-apple** [comment]
Hm, it seems like this is a source compat problem, since previously we let through certain switches that weren't actually exhaustive. What do you think we should do about this?

(One such case is [DLVM](https://github.com/dlvm-team/dlvm-core/), even though it's no longer being maintained. It still builds with [dlvm-patch.txt](https://github.com/apple/swift/files/2692418/dlvm-patch.txt), but the big switch [here](https://github.com/dlvm-team/dlvm-core/blob/master/Sources/DLVM/IR/Adjoints.swift#L56) is missing the case `.builtin(_, _)`.)

\[We're tracking this inside Apple as rdar://problem/46782440.]

**Comment by CodaFi** [comment]
As far as I'm concerned, it's a miscompile.  I didn't think the guarantees of source compatibility extended to those kinds of programs.

**Comment by jrose-apple** [comment]
Okay. It's still a source compat break, so we'll have to get sign-off from a core team member.

**Pull request merged**